pkg/client/kubernetes/cache/single_object.go

```diff
@@ -0,0 +1,189 @@
+		log:                       log,
+		config:                    config,
+		newCache:                  newCache,
+		opts:                      func() cache.Options { return opts },
+		objectKeyToCache:          make(map[client.ObjectKey]*objectCache),
+		clock:                     clock,
+		garbageCollectionInterval: garbageCollectionInterval,
+		maxIdleTime:               maxIdleTime,
+	}
+}
+
+func (s *singleObject) Start(ctx context.Context) error {
+	wait.Until(func() {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+
+		for key, objectCache := range s.objectKeyToCache {
+			lastAccessTime, ok := (objectCache.lastAccessTime.Load()).(*time.Time)
+			if !ok || lastAccessTime == nil || !s.clock.Now().After(lastAccessTime.Add(s.maxIdleTime)) {
+				continue
+			}
+
+			// close cache for object because it was not accessed for at least the max idle time
+			objectCache.cancel()
+			delete(s.objectKeyToCache, key)
+		}
+	}, s.garbageCollectionInterval, ctx.Done())
+	return nil
+}
+
+func (s *singleObject) WaitForCacheSync(_ context.Context) bool {
+	return true
+}
+
+func (s *singleObject) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	cache, err := s.getOrCreateCache(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	now := s.clock.Now()
+	cache.lastAccessTime.Store(&now)
+
+	return cache.cache.Get(ctx, key, obj, opts...)
+}
+
+func (s *singleObject) getOrCreateCache(ctx context.Context, key client.ObjectKey) (*objectCache, error) {
+	found, objectCache := func() (bool, *objectCache) {
+		s.lock.RLock()
+		defer s.lock.RUnlock()
+		objectCache, ok := s.objectKeyToCache[key]
+		return ok, objectCache
+	}()
+
+	if found {
+		s.log.V(1).Info("Cache found", "key", key)
+		return objectCache, nil
+	}
+
+	return s.createCache(ctx, key)
+}
+
+func (s *singleObject) createCache(ctx context.Context, key client.ObjectKey) (*objectCache, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// Cache could have been created since last read lock was set, so we need to check again if the cache is available.
+	if objectCache, ok := s.objectKeyToCache[key]; ok {
+		s.log.V(1).Info("Cache found before creation", "key", key)
+		return objectCache, nil
+	}
+
+	opts := s.opts()
+	opts.Namespace = key.Namespace
+	opts.DefaultSelector = cache.ObjectSelector{Field: fields.SelectorFromSet(fields.Set{metav1.ObjectNameField: key.Name})}
+	opts.SelectorsByObject = nil
+
+	s.log.V(1).Info("Creating new cache", "key", key)
+	c, err := s.newCache(s.config, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheCtx, cancel := context.WithCancel(ctx)
+
+	go func() {
+		s.log.V(1).Info("Starting new cache", "key", key)
+		if err := c.Start(cacheCtx); err != nil {
+			s.log.Error(err, "Cache failed to start", "key", key)
+		}
+	}()
+
+	if !c.WaitForCacheSync(ctx) {
+		cancel()
+		return nil, fmt.Errorf("failed waiting for cache to be synced")
+	}
+
+	s.objectKeyToCache[key] = &objectCache{
+		cache:  c,
+		cancel: cancel,
+	}
```

Comment:
Que - At what point in time `lastAccessTime` will be added. Should we not add  `lastAccessTime` as time.Now() here, cause cache would have been added because of some Get request.

Reply:
The time is updated here: https://github.com/gardener/gardener/pull/7632/files/8868d2c309711f68dbbe8d7cc2726f1620e22e1b#diff-3b51c37607a8ffa7507f2042aff828f7d441344e9ed4deb15248c88c0915e1c8R103-R104

However, we should indeed move it to `getOrCreateCache` since other functions like `GetInformer` or `IndexField` are also using the cache, it's not only `Get`.

Reply:
Thanks, I was concerned about if the `lastAccessTime` is not added then the object will never be removed from the cache.